src/index/coinstatsindex.cpp

```diff
@@ -0,0 +1,407 @@
+            return error("%s: previous block header belongs to unexpected block %s; expected %s",
+                         __func__, read_out.first.ToString(), expected_block_hash.ToString());
+        }
+
+        bool is_bip30_block = (pindex->nHeight==91722 && pindex->GetBlockHash() == uint256S("0x00000000000271a2dc26e7667f8419f2e15416dc6955e5a6c6cdf3f2574dd08e")) ||
+                              (pindex->nHeight==91812 && pindex->GetBlockHash() == uint256S("0x00000000000af0aed4792b1acee3d966af36cf5def14935db8de83d6f9306f2f"));
+        MuHash3072 undo_muhash;
+
+        // Add the new utxos created from the block
+        for (size_t i = 0; i < block.vtx.size(); ++i) {
+            const auto& tx = block.vtx.at(i);
+
+            // Skip duplicate txid coinbase transactions (BIP30).
+            if (is_bip30_block && tx->IsCoinBase()) {
+                continue;
+            }
+
+            for (size_t j = 0; j < tx->vout.size(); ++j) {
+                const CTxOut& out = tx->vout[j];
+                COutPoint outpoint = COutPoint(tx->GetHash(), j);
+                Coin coin = Coin(out, pindex->nHeight, tx->IsCoinBase());
+
+                // Skip unspendable coins
+                if (coin.out.scriptPubKey.IsUnspendable()) continue;
+
+                TruncatedSHA512Writer ss(SER_DISK, 0);
+                ss << outpoint;
+                ss << (uint32_t)(coin.nHeight * 2 + coin.fCoinBase);
+                ss << coin.out;
+                m_muhash *= MuHash3072(ss.GetHash().begin());
+
+                m_nTransactionOutputs++;
+                m_nTotalAmount += coin.out.nValue;
+                m_nBogoSize += 32 /* txid */ + 4 /* vout index */ + 4 /* height + coinbase */ + 8 /* amount */ +
+                               2 /* scriptPubKey len */ + coin.out.scriptPubKey.size() /* scriptPubKey */;
+            }
+
+            // The coinbase tx has no undo data since no former output is spent
+            if (i > 0) {
+                const auto& tx_undo = block_undo.vtxundo.at(i-1);
+
+                for (size_t j = 0; j < tx_undo.vprevout.size(); ++j) {
+                    Coin coin = tx_undo.vprevout[j];
+                    COutPoint outpoint = COutPoint(tx->vin[j].prevout.hash, tx->vin[j].prevout.n);
+
+                    TruncatedSHA512Writer ss(SER_DISK, 0);
+                    ss << outpoint;
+                    ss << (uint32_t)(coin.nHeight * 2 + coin.fCoinBase);
+                    ss << coin.out;
+                    undo_muhash *= MuHash3072(ss.GetHash().begin());
+
+                    m_nTransactionOutputs--;
+                    m_nTotalAmount -= coin.out.nValue;
+                    m_nBogoSize -= 32 /* txid */ + 4 /* vout index */ + 4 /* height + coinbase */ + 8 /* amount */ +
+                                   2 /* scriptPubKey len */ + coin.out.scriptPubKey.size() /* scriptPubKey */;
+                }
+            }
+        }
+
+        m_muhash /= undo_muhash;
```

Comment:
Why is this called `undo_muhash` and not `do_muhash`? And why is it divided?

Reply:
It is the intermediate muhash for all the utxos that need to be removed/undone. See   also my longer explanation for `do_muhash`.